{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Agapetus',\n",
       " 'Ameletus inopinatus',\n",
       " 'Amphinemura borealis',\n",
       " 'Baetis rhodani',\n",
       " 'Baetis vernus',\n",
       " 'Capnopsis schilleri',\n",
       " 'Diura',\n",
       " 'Elmis aenea',\n",
       " 'Ephemerella aurivillii',\n",
       " 'Ephemerella mucronata',\n",
       " 'Heptagenia sulphurea',\n",
       " 'Hydraena',\n",
       " 'Hydropsyche pellucidula',\n",
       " 'Hydropsyche saxonica',\n",
       " 'Hydropsyche siltalai',\n",
       " 'Isoperla',\n",
       " 'Kageronia fuscogrisea',\n",
       " 'Lepidostoma hirtum',\n",
       " 'Leptophlebia',\n",
       " 'Leuctra nigra',\n",
       " 'Leuctra',\n",
       " 'Limnius volckmari',\n",
       " 'Micrasema gelidum',\n",
       " 'Micrasema setiferum',\n",
       " 'Nemoura cinerea',\n",
       " 'Nemoura',\n",
       " 'Neureclipsis bimaculata',\n",
       " 'Oulimnius tuberculatus',\n",
       " 'Oxyethira',\n",
       " 'Plectrocnemia',\n",
       " 'Polycentropus flavomaculatus',\n",
       " 'Polycentropus irroratus',\n",
       " 'Protonemura',\n",
       " 'Rhyacophila nubila',\n",
       " 'Sialis',\n",
       " 'Silo pallipes',\n",
       " 'Simuliidae',\n",
       " 'Sphaerium',\n",
       " 'Taeniopteryx nebulosa']"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import yaml\n",
    "with open(\"../data/raw/bold_api/bold_names.yaml\", \"r\") as f:\n",
    "    config = yaml.safe_load(f)\n",
    "config[\"bold_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loaded 'Sialis'\n"
     ]
    }
   ],
   "source": [
    "from urllib.request import urlretrieve\n",
    "from pathlib import Path\n",
    "\n",
    "raw_folder = Path(\"../data/raw/bold_api\")\n",
    "out_folder = Path(\"../data/processed/bold_api\")\n",
    "\n",
    "def fetch_bold_sequences(taxon, out_fname):\n",
    "    url = (\"http://v3.boldsystems.org/index.php/API_Public/sequence?\"\n",
    "    f\"taxon={taxon.replace(' ', '%20')}\"\n",
    "        \"&marker=COI-5P\"\n",
    "        \"&format=xml\")\n",
    "\n",
    "    urlretrieve(url, filename=out_fname)\n",
    "    print(f\"loaded '{taxon}'\")\n",
    "\n",
    "taxon = \"Sialis\"\n",
    "out_fname = raw_folder / f\"{taxon.replace(' ', '_')}.fasta\"\n",
    "fetch_bold_sequences(taxon, out_fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loaded 'Agapetus'\n",
      "loaded 'Ameletus inopinatus'\n",
      "loaded 'Amphinemura borealis'\n",
      "loaded 'Baetis rhodani'\n",
      "loaded 'Baetis vernus'\n",
      "loaded 'Capnopsis schilleri'\n",
      "loaded 'Diura'\n",
      "loaded 'Elmis aenea'\n",
      "loaded 'Ephemerella aurivillii'\n",
      "loaded 'Ephemerella mucronata'\n",
      "loaded 'Heptagenia sulphurea'\n",
      "loaded 'Hydraena'\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[3], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m taxon \u001b[38;5;129;01min\u001b[39;00m config[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbold_name\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[0;32m      2\u001b[0m     out_fname \u001b[38;5;241m=\u001b[39m raw_folder \u001b[38;5;241m/\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtaxon\u001b[38;5;241m.\u001b[39mreplace(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.fasta\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m----> 3\u001b[0m     \u001b[43mfetch_bold_sequences\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtaxon\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mout_fname\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[1;32mIn[2], line 13\u001b[0m, in \u001b[0;36mfetch_bold_sequences\u001b[1;34m(taxon, out_fname)\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mfetch_bold_sequences\u001b[39m(taxon, out_fname):\n\u001b[0;32m      8\u001b[0m     url \u001b[38;5;241m=\u001b[39m (\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttp://v3.boldsystems.org/index.php/API_Public/sequence?\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m      9\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtaxon=\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtaxon\u001b[38;5;241m.\u001b[39mreplace(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;250m \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m%\u001b[39m\u001b[38;5;124m20\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m     10\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m&marker=COI-5P\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m     11\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m&format=xml\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m---> 13\u001b[0m     \u001b[43murlretrieve\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfilename\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mout_fname\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     14\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mloaded \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtaxon\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\E1007914\\AppData\\Local\\miniconda3\\envs\\taxonomist\\lib\\urllib\\request.py:268\u001b[0m, in \u001b[0;36murlretrieve\u001b[1;34m(url, filename, reporthook, data)\u001b[0m\n\u001b[0;32m    265\u001b[0m     reporthook(blocknum, bs, size)\n\u001b[0;32m    267\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m--> 268\u001b[0m     block \u001b[38;5;241m=\u001b[39m \u001b[43mfp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    269\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m block:\n\u001b[0;32m    270\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\E1007914\\AppData\\Local\\miniconda3\\envs\\taxonomist\\lib\\http\\client.py:463\u001b[0m, in \u001b[0;36mHTTPResponse.read\u001b[1;34m(self, amt)\u001b[0m\n\u001b[0;32m    460\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m amt \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    461\u001b[0m     \u001b[38;5;66;03m# Amount is given, implement using readinto\u001b[39;00m\n\u001b[0;32m    462\u001b[0m     b \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mbytearray\u001b[39m(amt)\n\u001b[1;32m--> 463\u001b[0m     n \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreadinto\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    464\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mmemoryview\u001b[39m(b)[:n]\u001b[38;5;241m.\u001b[39mtobytes()\n\u001b[0;32m    465\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    466\u001b[0m     \u001b[38;5;66;03m# Amount is not given (unbounded read) so we must check self.length\u001b[39;00m\n\u001b[0;32m    467\u001b[0m     \u001b[38;5;66;03m# and self.chunked\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\E1007914\\AppData\\Local\\miniconda3\\envs\\taxonomist\\lib\\http\\client.py:497\u001b[0m, in \u001b[0;36mHTTPResponse.readinto\u001b[1;34m(self, b)\u001b[0m\n\u001b[0;32m    494\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n\u001b[0;32m    496\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchunked:\n\u001b[1;32m--> 497\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_readinto_chunked\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    499\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlength \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    500\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(b) \u001b[38;5;241m>\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlength:\n\u001b[0;32m    501\u001b[0m         \u001b[38;5;66;03m# clip the read to the \"end of response\"\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\E1007914\\AppData\\Local\\miniconda3\\envs\\taxonomist\\lib\\http\\client.py:592\u001b[0m, in \u001b[0;36mHTTPResponse._readinto_chunked\u001b[1;34m(self, b)\u001b[0m\n\u001b[0;32m    590\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m    591\u001b[0m     \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m--> 592\u001b[0m         chunk_left \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_chunk_left\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    593\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m chunk_left \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    594\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m total_bytes\n",
      "File \u001b[1;32mc:\\Users\\E1007914\\AppData\\Local\\miniconda3\\envs\\taxonomist\\lib\\http\\client.py:560\u001b[0m, in \u001b[0;36mHTTPResponse._get_chunk_left\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    558\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_safe_read(\u001b[38;5;241m2\u001b[39m)  \u001b[38;5;66;03m# toss the CRLF at the end of the chunk\u001b[39;00m\n\u001b[0;32m    559\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 560\u001b[0m     chunk_left \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_read_next_chunk_size\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    561\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m:\n\u001b[0;32m    562\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m IncompleteRead(\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\E1007914\\AppData\\Local\\miniconda3\\envs\\taxonomist\\lib\\http\\client.py:520\u001b[0m, in \u001b[0;36mHTTPResponse._read_next_chunk_size\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    518\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_read_next_chunk_size\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m    519\u001b[0m     \u001b[38;5;66;03m# Read the next chunk size from the file\u001b[39;00m\n\u001b[1;32m--> 520\u001b[0m     line \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreadline\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_MAXLINE\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m    521\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(line) \u001b[38;5;241m>\u001b[39m _MAXLINE:\n\u001b[0;32m    522\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m LineTooLong(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mchunk size\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\E1007914\\AppData\\Local\\miniconda3\\envs\\taxonomist\\lib\\socket.py:704\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[1;34m(self, b)\u001b[0m\n\u001b[0;32m    702\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m    703\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 704\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv_into\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    705\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n\u001b[0;32m    706\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_timeout_occurred \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "for taxon in config[\"bold_name\"]:\n",
    "    out_fname = raw_folder / f\"{taxon.replace(' ', '_')}.fasta\"\n",
    "    fetch_bold_sequences(taxon, out_fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from random import sample\n",
    "from io import StringIO\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def preprocess_fasta(input):\n",
    "    folder = input.parent\n",
    "    records = [r for r in SeqIO.parse(input, \"fasta\")]\n",
    "    print(\"Original length:\", len(records))\n",
    "\n",
    "    # Find the lengths of sequences and filter too long and short sequences\n",
    "    lengths = np.asarray([len(r.seq) for r in records])\n",
    "    filter_ind = np.logical_and((lengths <= 658), (lengths >= 600))\n",
    "    records_subset0 = [records[i] for i in np.where(filter_ind)[0]]\n",
    "    print(\"Sequences between 600 and 658:\", len(records_subset0))\n",
    "\n",
    "    # Remove duplicate sequences\n",
    "    seen = set()\n",
    "    records_subset = []\n",
    "    for r in records_subset0:\n",
    "        if r.seq not in seen:\n",
    "            seen.add(r.seq)\n",
    "            records_subset.append(r)\n",
    "\n",
    "    print(\"After duplicates removed:\", len(records_subset))\n",
    "\n",
    "    # Cap the maximum number of sequences to 500\n",
    "    if len(records_subset) > 200:\n",
    "        records_subset = sample(records_subset, 200)\n",
    "        print(\"Capped to 200:\", len(records_subset))\n",
    "\n",
    "    subset_name = raw_folder / f\"{input.stem}_subset.fasta\"\n",
    "    with open(subset_name, \"w\") as output_handle:\n",
    "        SeqIO.write(records_subset, output_handle, \"fasta\")\n",
    "    print(subset_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original length: 652\n",
      "Sequences between 600 and 658: 596\n",
      "After duplicates removed: 359\n",
      "Capped to 200: 200\n",
      "..\\data\\raw\\bold_api\\Agapetus_subset.fasta\n",
      "Original length: 42\n",
      "Sequences between 600 and 658: 41\n",
      "After duplicates removed: 23\n",
      "..\\data\\raw\\bold_api\\Ameletus_inopinatus_subset.fasta\n",
      "Original length: 14\n",
      "Sequences between 600 and 658: 14\n",
      "After duplicates removed: 7\n",
      "..\\data\\raw\\bold_api\\Amphinemura_borealis_subset.fasta\n",
      "Original length: 588\n",
      "Sequences between 600 and 658: 413\n",
      "After duplicates removed: 341\n",
      "Capped to 200: 200\n",
      "..\\data\\raw\\bold_api\\Baetis_rhodani_subset.fasta\n",
      "Original length: 159\n",
      "Sequences between 600 and 658: 102\n",
      "After duplicates removed: 55\n",
      "..\\data\\raw\\bold_api\\Baetis_vernus_subset.fasta\n",
      "Original length: 13\n",
      "Sequences between 600 and 658: 10\n",
      "After duplicates removed: 9\n",
      "..\\data\\raw\\bold_api\\Capnopsis_schilleri_subset.fasta\n",
      "Original length: 74\n",
      "Sequences between 600 and 658: 57\n",
      "After duplicates removed: 37\n",
      "..\\data\\raw\\bold_api\\Diura_subset.fasta\n",
      "Original length: 331\n",
      "Sequences between 600 and 658: 313\n",
      "After duplicates removed: 73\n",
      "..\\data\\raw\\bold_api\\Elmis_aenea_subset.fasta\n",
      "Original length: 39\n",
      "Sequences between 600 and 658: 38\n",
      "After duplicates removed: 29\n",
      "..\\data\\raw\\bold_api\\Ephemerella_aurivillii_subset.fasta\n",
      "Original length: 29\n",
      "Sequences between 600 and 658: 23\n",
      "After duplicates removed: 17\n",
      "..\\data\\raw\\bold_api\\Ephemerella_mucronata_subset.fasta\n",
      "Original length: 19\n",
      "Sequences between 600 and 658: 18\n",
      "After duplicates removed: 17\n",
      "..\\data\\raw\\bold_api\\Heptagenia_sulphurea_subset.fasta\n",
      "Original length: 300\n",
      "Sequences between 600 and 658: 242\n",
      "After duplicates removed: 174\n",
      "..\\data\\raw\\bold_api\\Hydraena_subset.fasta\n",
      "Original length: 12\n",
      "Sequences between 600 and 658: 11\n",
      "After duplicates removed: 8\n",
      "..\\data\\raw\\bold_api\\Hydropsyche_pellucidula_subset.fasta\n",
      "Original length: 86\n",
      "Sequences between 600 and 658: 78\n",
      "After duplicates removed: 19\n",
      "..\\data\\raw\\bold_api\\Hydropsyche_saxonica_subset.fasta\n",
      "Original length: 105\n",
      "Sequences between 600 and 658: 96\n",
      "After duplicates removed: 23\n",
      "..\\data\\raw\\bold_api\\Hydropsyche_siltalai_subset.fasta\n",
      "Original length: 1237\n",
      "Sequences between 600 and 658: 965\n",
      "After duplicates removed: 709\n",
      "Capped to 200: 200\n",
      "..\\data\\raw\\bold_api\\Isoperla_subset.fasta\n",
      "Original length: 16\n",
      "Sequences between 600 and 658: 16\n",
      "After duplicates removed: 9\n",
      "..\\data\\raw\\bold_api\\Kageronia_fuscogrisea_subset.fasta\n",
      "Original length: 93\n",
      "Sequences between 600 and 658: 85\n",
      "After duplicates removed: 64\n",
      "..\\data\\raw\\bold_api\\Lepidostoma_hirtum_subset.fasta\n",
      "Original length: 270\n",
      "Sequences between 600 and 658: 220\n",
      "After duplicates removed: 145\n",
      "..\\data\\raw\\bold_api\\Leptophlebia_subset.fasta\n",
      "Original length: 75\n",
      "Sequences between 600 and 658: 66\n",
      "After duplicates removed: 33\n",
      "..\\data\\raw\\bold_api\\Leuctra_nigra_subset.fasta\n",
      "Original length: 1819\n",
      "Sequences between 600 and 658: 1499\n",
      "After duplicates removed: 944\n",
      "Capped to 200: 200\n",
      "..\\data\\raw\\bold_api\\Leuctra_subset.fasta\n",
      "Original length: 34\n",
      "Sequences between 600 and 658: 33\n",
      "After duplicates removed: 18\n",
      "..\\data\\raw\\bold_api\\Limnius_volckmari_subset.fasta\n",
      "Original length: 5\n",
      "Sequences between 600 and 658: 4\n",
      "After duplicates removed: 4\n",
      "..\\data\\raw\\bold_api\\Micrasema_gelidum_subset.fasta\n",
      "Original length: 34\n",
      "Sequences between 600 and 658: 25\n",
      "After duplicates removed: 19\n",
      "..\\data\\raw\\bold_api\\Micrasema_setiferum_subset.fasta\n",
      "Original length: 101\n",
      "Sequences between 600 and 658: 82\n",
      "After duplicates removed: 45\n",
      "..\\data\\raw\\bold_api\\Nemoura_cinerea_subset.fasta\n",
      "Original length: 848\n",
      "Sequences between 600 and 658: 772\n",
      "After duplicates removed: 481\n",
      "Capped to 200: 200\n",
      "..\\data\\raw\\bold_api\\Nemoura_subset.fasta\n",
      "Original length: 48\n",
      "Sequences between 600 and 658: 35\n",
      "After duplicates removed: 17\n",
      "..\\data\\raw\\bold_api\\Neureclipsis_bimaculata_subset.fasta\n",
      "Original length: 16\n",
      "Sequences between 600 and 658: 16\n",
      "After duplicates removed: 5\n",
      "..\\data\\raw\\bold_api\\Oulimnius_tuberculatus_subset.fasta\n",
      "Original length: 562\n",
      "Sequences between 600 and 658: 398\n",
      "After duplicates removed: 329\n",
      "Capped to 200: 200\n",
      "..\\data\\raw\\bold_api\\Oxyethira_subset.fasta\n",
      "Original length: 616\n",
      "Sequences between 600 and 658: 578\n",
      "After duplicates removed: 343\n",
      "Capped to 200: 200\n",
      "..\\data\\raw\\bold_api\\Plectrocnemia_subset.fasta\n",
      "Original length: 55\n",
      "Sequences between 600 and 658: 53\n",
      "After duplicates removed: 25\n",
      "..\\data\\raw\\bold_api\\Polycentropus_flavomaculatus_subset.fasta\n",
      "Original length: 28\n",
      "Sequences between 600 and 658: 20\n",
      "After duplicates removed: 11\n",
      "..\\data\\raw\\bold_api\\Polycentropus_irroratus_subset.fasta\n",
      "Original length: 286\n",
      "Sequences between 600 and 658: 230\n",
      "After duplicates removed: 164\n",
      "..\\data\\raw\\bold_api\\Protonemura_subset.fasta\n",
      "Original length: 67\n",
      "Sequences between 600 and 658: 65\n",
      "After duplicates removed: 35\n",
      "..\\data\\raw\\bold_api\\Rhyacophila_nubila_subset.fasta\n",
      "Original length: 155\n",
      "Sequences between 600 and 658: 110\n",
      "After duplicates removed: 86\n",
      "..\\data\\raw\\bold_api\\Sialis_subset.fasta\n",
      "Original length: 61\n",
      "Sequences between 600 and 658: 53\n",
      "After duplicates removed: 37\n",
      "..\\data\\raw\\bold_api\\Silo_pallipes_subset.fasta\n",
      "Original length: 19514\n",
      "Sequences between 600 and 658: 15667\n",
      "After duplicates removed: 11079\n",
      "Capped to 200: 200\n",
      "..\\data\\raw\\bold_api\\Simuliidae_subset.fasta\n",
      "Original length: 124\n",
      "Sequences between 600 and 658: 94\n",
      "After duplicates removed: 70\n",
      "..\\data\\raw\\bold_api\\Sphaerium_subset.fasta\n",
      "Original length: 22\n",
      "Sequences between 600 and 658: 21\n",
      "After duplicates removed: 14\n",
      "..\\data\\raw\\bold_api\\Taeniopteryx_nebulosa_subset.fasta\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "for taxon in config[\"bold_name\"]:\n",
    "    out_fname = raw_folder / f\"{taxon.replace(' ', '_')}.fasta\"\n",
    "    preprocess_fasta(out_fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_records(filename):\n",
    "    return [r for r in SeqIO.parse(filename, \"fasta\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Aligning Agapetus, 200 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Agapetus_subset.afa\n",
      "Aligning Ameletus inopinatus, 23 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Ameletus_inopinatus_subset.afa\n",
      "Aligning Amphinemura borealis, 7 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Amphinemura_borealis_subset.afa\n",
      "Aligning Baetis rhodani, 200 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Baetis_rhodani_subset.afa\n",
      "Aligning Baetis vernus, 55 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Baetis_vernus_subset.afa\n",
      "Aligning Capnopsis schilleri, 9 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Capnopsis_schilleri_subset.afa\n",
      "Aligning Diura, 37 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Diura_subset.afa\n",
      "Aligning Elmis aenea, 73 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Elmis_aenea_subset.afa\n",
      "Aligning Ephemerella aurivillii, 29 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Ephemerella_aurivillii_subset.afa\n",
      "Aligning Ephemerella mucronata, 17 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Ephemerella_mucronata_subset.afa\n",
      "Aligning Heptagenia sulphurea, 17 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Heptagenia_sulphurea_subset.afa\n",
      "Aligning Hydraena, 174 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Hydraena_subset.afa\n",
      "Aligning Hydropsyche pellucidula, 8 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Hydropsyche_pellucidula_subset.afa\n",
      "Aligning Hydropsyche saxonica, 19 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Hydropsyche_saxonica_subset.afa\n",
      "Aligning Hydropsyche siltalai, 23 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Hydropsyche_siltalai_subset.afa\n",
      "Aligning Isoperla, 200 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Isoperla_subset.afa\n",
      "Aligning Kageronia fuscogrisea, 9 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Kageronia_fuscogrisea_subset.afa\n",
      "Aligning Lepidostoma hirtum, 64 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Lepidostoma_hirtum_subset.afa\n",
      "Aligning Leptophlebia, 145 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Leptophlebia_subset.afa\n",
      "Aligning Leuctra nigra, 33 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Leuctra_nigra_subset.afa\n",
      "Aligning Leuctra, 200 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Leuctra_subset.afa\n",
      "Aligning Limnius volckmari, 18 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Limnius_volckmari_subset.afa\n",
      "Aligning Micrasema gelidum, 4 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Micrasema_gelidum_subset.afa\n",
      "Aligning Micrasema setiferum, 19 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Micrasema_setiferum_subset.afa\n",
      "Aligning Nemoura cinerea, 45 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Nemoura_cinerea_subset.afa\n",
      "Aligning Nemoura, 200 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Nemoura_subset.afa\n",
      "Aligning Neureclipsis bimaculata, 17 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Neureclipsis_bimaculata_subset.afa\n",
      "Aligning Oulimnius tuberculatus, 5 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Oulimnius_tuberculatus_subset.afa\n",
      "Aligning Oxyethira, 200 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Oxyethira_subset.afa\n",
      "Aligning Plectrocnemia, 200 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Plectrocnemia_subset.afa\n",
      "Aligning Polycentropus flavomaculatus, 25 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Polycentropus_flavomaculatus_subset.afa\n",
      "Aligning Polycentropus irroratus, 11 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Polycentropus_irroratus_subset.afa\n",
      "Aligning Protonemura, 164 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Protonemura_subset.afa\n",
      "Aligning Rhyacophila nubila, 35 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Rhyacophila_nubila_subset.afa\n",
      "Aligning Sialis, 86 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Sialis_subset.afa\n",
      "Aligning Silo pallipes, 37 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Silo_pallipes_subset.afa\n",
      "Aligning Simuliidae, 200 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Simuliidae_subset.afa\n",
      "Aligning Sphaerium, 70 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Sphaerium_subset.afa\n",
      "Aligning Taeniopteryx nebulosa, 14 records\n",
      "Aligned to file ..\\data\\processed\\bold_api\\Taeniopteryx_nebulosa_subset.afa\n"
     ]
    }
   ],
   "source": [
    "for taxon in config[\"bold_name\"]:\n",
    "    subset_name = raw_folder / f\"{taxon.replace(' ', '_')}_subset.fasta\"\n",
    "    align_name = out_folder / f\"{subset_name.stem}.afa\"\n",
    "    records = read_records(subset_name)\n",
    "    print(f\"Aligning {taxon}, {len(records)} records\")\n",
    "    subprocess.run([r\"..\\muscle.exe\",\n",
    "                        \"-super5\", str(subset_name), \n",
    "                        \"-output\", str(align_name)]\n",
    "                        )\n",
    "    print(f\"Aligned to file {align_name}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "records = []\n",
    "for taxon in config[\"bold_name\"]:\n",
    "    subset_name = raw_folder / f\"{taxon.replace(' ', '_')}_subset.fasta\"\n",
    "    align_name = out_folder / f\"{subset_name.stem}.afa\"\n",
    "    for r in read_records(align_name):\n",
    "        records.append(r)\n",
    "\n",
    "with open(\"../data/processed/fb2.fasta\", \"w\") as output_handle:\n",
    "    SeqIO.write(records, output_handle, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "taxonomist",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
